File: Components/DevExpressVCL/Demos/ExpressPivotGrid/Delphi/CompactLayoutDemo/CompactLayoutMain.pas
unit CompactLayoutMain;

{$I cxVer.inc}

interface

uses
  Windows, Messages, Variants, Forms, Dialogs,
  SysUtils, Menus, StdCtrls, Controls, cxLookAndFeels, DemoBasicMain,
  cxControls, cxCustomPivotGrid, cxDBPivotGrid, DemoBasicDM, cxStyles,
  cxCustomData, cxGraphics, cxFilter, cxData, cxDataStorage, cxEdit, DB, cxDBData,
  Classes, cxClasses, cxLookAndFeelPainters, cxSplitter;

type
  TfrmCompactLayout = class(TfrmDemoBasicMain)
    DBPivotGrid: TcxDBPivotGrid;
    pgfPurchaseDate: TcxDBPivotGridField;
    pgfPaymentType: TcxDBPivotGridField;
    pgfQuantity: TcxDBPivotGridField;
    pgfCarName: TcxDBPivotGridField;
    pgfUnitPrice: TcxDBPivotGridField;
    pgfCompanyName: TcxDBPivotGridField;
    pgfPaymentAmount: TcxDBPivotGridField;
    miCustomization: TMenuItem;
    cxSplitter1: TcxSplitter;
    procedure FormCreate(Sender: TObject);
    procedure pgfPaymentTypeGetGroupImageIndex(Sender: TcxPivotGridField;
      const AItem: TcxPivotGridViewDataItem; var AImageIndex: Integer;
      var AImageAlignHorz: TAlignment;
      var AImageAlignVert: TcxAlignmentVert);
    procedure miCustomizationClick(Sender: TObject);
    procedure DBPivotGridCustomization(Sender: TObject);
  private
  protected
    function GetPivotGrid: TcxCustomPivotGrid; override;
  public
    { Public declarations }
  end;

var
  frmCompactLayout: TfrmCompactLayout;

implementation

{$R *.dfm}

uses
  DemoUtils;

function TfrmCompactLayout.GetPivotGrid: TcxCustomPivotGrid;
begin
  Result := DBPivotGrid;
end;

procedure TfrmCompactLayout.FormCreate(Sender: TObject);
begin
  inherited;
  PivotGrid.ApplyBestFit;
  PivotGrid.Customization.Site := Self;
  pgfPurchaseDate.ExpandAll;
  pgfCarName.ExpandAll;
  pgfCompanyName.ExpandAll;
  PivotGrid.Customization.Visible := True;
end;

procedure TfrmCompactLayout.pgfPaymentTypeGetGroupImageIndex(
  Sender: TcxPivotGridField; const AItem: TcxPivotGridViewDataItem;
  var AImageIndex: Integer; var AImageAlignHorz: TAlignment;
  var AImageAlignVert: TcxAlignmentVert);
var
  Card: string;
begin
  Card := VarToStr(AItem.Value);
  if SameText(Card, 'AmEx') then
    AImageIndex := 1
  else
    if SameText(Card, 'Cash') then
      AImageIndex := 0
    else
      if SameText(Card, 'Master') then
        AImageIndex := 2
      else
        if SameText(Card, 'Visa') then
          AImageIndex := 3;
end;

procedure TfrmCompactLayout.miCustomizationClick(Sender: TObject);
begin
  inherited;
  PivotGrid.Customization.Visible := not PivotGrid.Customization.Visible;
end;

procedure TfrmCompactLayout.DBPivotGridCustomization(Sender: TObject);
begin
  inherited;
  dxDemoMenuItemSetChecked(dxDemoFindMenuItem(mmMain, 'miCustomization'), PivotGrid.Customization.Visible);
  if PivotGrid.Customization.Visible then
    PivotGrid.Customization.Form.Align := alLeft;
  cxSplitter1.Control := PivotGrid.Customization.Form;
  cxSplitter1.Visible := PivotGrid.Customization.Visible;
end;

end.

File: Components/DevExpressVCL/Demos/ExpressPivotGrid/CBuilder/CompactLayoutDemo/CompactLayoutMain.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "CompactLayoutMain.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "cxLookAndFeels"
#pragma link "DemoBasicMain"
#pragma link "cxControls"
#pragma link "cxCustomPivotGrid"
#pragma link "cxDBPivotGrid"
#pragma link "cxClasses"
#pragma link "cxCustomData"
#pragma link "cxEdit"
#pragma link "cxGraphics"
#pragma link "cxLookAndFeelPainters"
#pragma link "cxStyles"
#pragma resource "*.dfm"
TfrmCompactLayout *frmCompactLayout;
//---------------------------------------------------------------------------
__fastcall TfrmCompactLayout::TfrmCompactLayout(TComponent* Owner)
		: TfrmDemoBasicMain(Owner)
{
}

void __fastcall TfrmCompactLayout::FormCreate(TObject* Sender)
{
  TfrmDemoBasicMain::FormCreate(Sender);
  PivotGrid()->ApplyBestFit();
  PivotGrid()->Customization->Site = this;
  pgfPurchaseDate->ExpandAll();
  pgfCarName->ExpandAll();
  pgfCompanyName->ExpandAll();
  PivotGrid()->Customization->Visible = true;
}

TcxCustomPivotGrid* __fastcall TfrmCompactLayout::PivotGrid()
{
  return DBPivotGrid;
}

//---------------------------------------------------------------------------
void __fastcall TfrmCompactLayout::pgfPaymentTypeGetGroupImageIndex(
	  TcxPivotGridField *Sender, const TcxPivotGridViewDataItem *AItem,
	  int &AImageIndex, TAlignment &AImageAlignHorz,
	  TcxAlignmentVert &AImageAlignVert)
{
  AnsiString Card = ((TcxPivotGridViewDataItem *)AItem)->Value;
  if (SameText(Card, "Cash")) AImageIndex = 0;
  else if (SameText(Card, "AmEx")) AImageIndex = 1;
	   else if (SameText(Card, "Master")) AImageIndex = 2;
			else if (SameText(Card, "Visa")) AImageIndex = 3;
}
//---------------------------------------------------------------------------
void __fastcall TfrmCompactLayout::miCustomizationClick(TObject * Sender)
{
  PivotGrid()->Customization->Visible = !PivotGrid()->Customization->Visible;
}
//---------------------------------------------------------------------------
void __fastcall TfrmCompactLayout::DBPivotGridCustomization(TObject * Sender)
{
  miCustomization->Checked = PivotGrid()->Customization->Visible;
  if (PivotGrid()->Customization->Visible)
    PivotGrid()->Customization->Form->Align = alLeft;
  cxSplitter1->Control = PivotGrid()->Customization->Form;
  cxSplitter1->Visible = PivotGrid()->Customization->Visible;
}
//---------------------------------------------------------------------------
